README.md
spectris
========

A music and sound spectrum visualizer using OpenGL and BASS library


CHANGELOG
v0.1.0-alpha

	** Spectrum now works!
	** Updated motion library
	++ Added vector library

v0.0.0-alpha

	++ Added basic spectrum using BASS library


src/main.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include <GL/glew.h>
#include <GL/glut.h>

#define VECTOR_IMPLEMENTATION
#include "../include/vector.h"
#define MOTION_IMPLEMENTATION
#include "../include/motion.h"
#include "../include/utility.h"
#include "../include/bass.h"

// C does not support boolean
#define true 1
#define false 0

// Ligth parameters
GLfloat lightPos[] = {-1, 4.5, 16, 1};
GLfloat lightAmbient[] = {0.5, 0.5, 0.5, 1};
GLfloat lightDiffuse[] = {0.8, 0.8, 0.8, 1};
GLfloat lightSpecular[] = {0.4, 0.4, 0.4, 1};

int soundStream = 0;
int skipToDrop = false;
int dropPos = 95;

// This will draw a 100 by 100 plane, the camera being always in the middle of it. Gives the impression of an infinite world
void drawGround() {
	int centerX = (int) mot_GetEyePos().x, centerY = (int) mot_GetEyePos().z;
	int i, j;
	glEnable(GL_TEXTURE_2D);
	glPushMatrix();
		glTranslatef(0, -0.01, 0);
		glBegin(GL_QUADS);
		glNormal3f(0, 1, 0);
			for (i = 0; i < 100; i++) {
				for (j = 0; j < 100; j++) {
					glTexCoord2f(1, 0);
					glVertex3f(centerX + i - 50, 0, centerY + j - 50);
					glTexCoord2f(1, 1);
					glVertex3f(centerX + i - 50 + 1, 0, centerY + j - 50);
					glTexCoord2f(0, 1);
					glVertex3f(centerX + i - 50 + 1, 0, centerY + j - 50 + 1);
					glTexCoord2f(0, 0);
					glVertex3f(centerX + i - 50, 0, centerY + j - 50 + 1);
				}
			}
		glEnd();
	glPopMatrix();
	glDisable(GL_TEXTURE_2D);
}

void drawSpectrum(void) {
	// FFT size (2 ^ n)
	int n = 10;
	// Number of bands
	int bandNum = 32;

	// Get FFT
	int N = (int)pow(2, n);
	float fft[N];
	if (BASS_ChannelGetData(soundStream, &fft, BASS_DATA_FFT2048) == -1) return;

	// Spits the buffer in bandNum bands and computes the magnitude for each one
	int i, j;
	for (i = 0; i < bandNum; i++) {
		float peak = 0;
		int p1 = (int)pow(2, i * n / (float)(bandNum - 1));
		if (p1 > N - 1) p1 = N - 1;
		if (p1 <= 0) p1 = 1;

		int p0 = (int)pow(2, (i - 1) * n / (float)(bandNum - 1));
		if (p0 > N - 1) p0 = N - 1;
		if (p0 < 0) p0 = 0;

		for (j = p0; j < p1; j++)
		{
		    if (peak < fft[j + 1]) peak = fft[j + 1];
		}
	
		// Draw band
		float scale = sqrt(peak) * 30;
		float width = 0.5;
		float spacing = 0.5;

		glPushMatrix();
			glTranslatef(0, scale * width / 2, i * (width + spacing));
			glScalef(1, scale, 1);
			glColor3f(0, 0.2, 1);
			glutSolidCube(width);
		glPopMatrix();
	}


	// // Failed attempt 2
	// for (i = 0; i < bandNum; i++) {
	// 	float peak = 0;
	// 	int p = (int)pow(2, i * 10.0 / (bandNum - 1));
	// 	if (p > 1023) p = 1023;
	// 	if (p <= 0) p = 1;
	// 	for (j = 0; j < p; j++)
	// 	{
	// 	    if (peak < fft[j + 1]) peak = fft[j + 1];
	// 	}
	// 	float y = sqrt(peak);
	
	// 	// Draw band
	// 	float scale = y * 30;
	// 	glPushMatrix();
	// 		glTranslatef(0, scale * 0.25, i);
	// 		glScalef(1, scale, 1);
	// 		glColor3f(0, 0.2, 1);
	// 		glutSolidCube(0.5);
	// 	glPopMatrix();
	// }


	// // Failed attempt 1
	// GLdouble mag;
	// for (i = 0; i < bandNum; i++) {
	// 	mag = 0;
	// 	for (j = 0; j < N / bandNum; j++) {
	// 			mag += fft[i * (N / bandNum) + j + 1];
	// 	}
	// 	// Average
	// 	mag /= N / bandNum;

	// 	// Draw band

	// 	float scaleCoef = 0.05 * (-20) * log10(mag);
	// 	glPushMatrix();
	// 		glTranslatef(0, scaleCoef * 0.25, i);
	// 		glScalef(1, scaleCoef, 1);
	// 		glColor3f(0, 0.2, 1);
	// 		glutSolidCube(0.5);
	// 	glPopMatrix();
	// }

	// Reset color
	glColor3f(0.5, 0.5, 0.5);
}

void display(void)
{
	// Clear the color buffer, restore the background
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
	// Load the identity matrix, clear all the previous transformations
	glLoadIdentity();
	// Set up the camera
	mot_MoveCamera();
	mot_SetCamera();
	// Set light position
	glLightfv(GL_LIGHT0, GL_POSITION, lightPos);

	// Draws ground
	// drawGround();

	// Draw spectrum
	drawSpectrum();

	// Swap buffers in GPU
	glutSwapBuffers();
}

void reshape(int width, int height)
{
	// Set the viewport to the full window size
	glViewport(0, 0, (GLsizei) width, (GLsizei) height);
	// Load the projection matrix
	glMatrixMode(GL_PROJECTION);
	// Clear all the transformations on the projection matrix
	glLoadIdentity();
	// Set the perspective according to the window size
	gluPerspective(70, (GLdouble) width / (GLdouble) height, 0.1, 60000);
	// Load back the modelview matrix
	glMatrixMode(GL_MODELVIEW);
}

void initialize(void)
{
	// Set the background to light gray
	glClearColor(0.8, 0.8, 0.8, 1);
	// Enables depth test
	glEnable(GL_DEPTH_TEST);
	// Disable color material
	glEnable(GL_COLOR_MATERIAL);
	// Enable normalize 
	glEnable(GL_NORMALIZE);
	// Enable Blend and transparency
	glEnable(GL_BLEND);
	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

	// Sets the light
	glLightfv(GL_LIGHT0, GL_AMBIENT, lightAmbient);
	glLightfv(GL_LIGHT0, GL_DIFFUSE, lightDiffuse);
	glLightfv(GL_LIGHT0, GL_SPECULAR, lightSpecular);
	// Enables lighting
	glEnable(GL_LIGHTING);
	// Enables the light
	glEnable(GL_LIGHT0);

	glShadeModel(GL_SMOOTH);
}

void tick(int value)
{	
	// Calls the display() function
	glutPostRedisplay();

	// Waits 10 ms
	glutTimerFunc(10, tick, 0);
}

// Called when the song ends
void replayMusic(HSYNC handle, DWORD channel, DWORD data, void *user) {
	// Replays the song
	BASS_ChannelPlay(channel, TRUE);
}

void bass(void) {
	// Initializes the library
	BASS_Init(-1, 44100, 0, 0, NULL);
	// Creates a stream from a file (the stream is some sort of ID)
	soundStream = BASS_StreamCreateFile(FALSE, "data/sound/Tritonal  Paris Blohm ft Sterling Fox - Colors Culture Code Remix.mp3", 0, 0, 0);
	// Adds an "Event Listener" (it is called sync) to detect when the song ends
	BASS_ChannelSetSync(soundStream, BASS_SYNC_MIXTIME | BASS_SYNC_END, 0, replayMusic, 0);
	// Plays the stream
	BASS_ChannelPlay(soundStream, TRUE);
	if (skipToDrop) {
		BASS_ChannelSetPosition(soundStream, BASS_ChannelSeconds2Bytes(soundStream, dropPos), BASS_POS_BYTE);	
	}
}


int main(int argc, char *argv[])
{
	printf("===================================================\n");
	glutInit(&argc, argv);
	glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB | GLUT_DEPTH);
	glutInitWindowSize(800, 600);
	glutCreateWindow("Spectris");
	glewInit();

	// Event listeners
	glutDisplayFunc(display);
	glutReshapeFunc(reshape);
	
	initialize();
	mot_Init(1.0 / 80);
	mot_TeleportCamera(-20, mot_GetConstant(MOT_EYE_HEIGHT), 15);
	bass();

	// Starts main timer
	glutTimerFunc(10, tick, 0);

	glutMainLoop();
	return 0;
}


src/utility.c
#include <stdio.h>
#include <stdlib.h>

#include <GL/glew.h>

#include "../include/utility.h"

// Reads all the content of file, creates a shader, adds the source and compiles it. Returns shader's id
GLuint loadShaderFromFile(char * path, GLenum type) {
	// Creates Shader
	GLuint id = glCreateShader(type);

	// Opens the file
	FILE * fp = fopen(path, "rb");
	if (fp == NULL) {
		printf("ERROR: Failed to open \"%s\".\n", path);
		return id;
	}

	// Gets file length
	int length;
	fseek(fp, 0, SEEK_END);
	length = ftell(fp);
	fseek(fp, 0, SEEK_SET);

	// Allocates memory and loads file's content
	char * code;
	code = (char *) malloc(length * sizeof(char));
	fread(code, length * sizeof(char), 1, fp);

	// Close file
	fclose(fp);

	// Adds shader source and compiles it
	glShaderSource(id, 1, (const GLchar **) &code, NULL);
	glCompileShader(id);

	// Warns the user if the compilation failed
	GLint success = 0;
	glGetShaderiv(id, GL_COMPILE_STATUS, &success);
	if (success == GL_FALSE) {
		printf("WARNING: Shader at \"%s\" didn't compile\n", path);
	}
	else {
		//printf("Shader at \"%s\" compiled successfully\n", path);
	}
	return id;
}

// Creates a program and links two shaders. Returns the program's id.
GLuint createShadersProgram(GLuint id1, GLuint id2) {
	GLuint id = glCreateProgram();
	glAttachShader(id, id1);
	glAttachShader(id, id2);
	glLinkProgram(id);
	return id;
}

// Loads two Shaders from files
void loadShaders(char * path1, GLenum type1, char * path2, GLenum type2) {
	GLuint id1 = loadShaderFromFile(path1, type1);
	GLuint id2 = loadShaderFromFile(path2, type2);
	GLuint programID = createShadersProgram(id1, id2);
	glUseProgram(programID);
}

src/motion.c
/* 
	This handles camera movement
*/

#include <GL/glut.h>
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include "../include/utility.h"
#include "../include/motion.h"

// Camerea pitch and yaw
GLdouble cameraYaw = 0, cameraPitch = 0;
// Vectors
vector eye, target, cameraForce;
// Key state: true if the key is pressed and false if not
int keyStates[256];
int isJumping = false;
int isSprinting = false;

void jumpFunc(void) {
	// The player jumps only if the spacebar is pressed and if he is not jumping
	if (!isJumping && keyStates[32]) {
		isJumping = true;

		// Makes a push vector based on input and direction (this allows sprint jumping)
		vector direction = substractv(createv(target.x, EYE_HEIGHT, target.z), eye);
		normalizev(&direction);
		vector push = createv(0, 0, 0);
		if (keyStates['w']) {
			push = addv(push, direction);
		}
		if (keyStates['s']) {
			push = substractv(push, direction);
		}
		if (keyStates['a']) {
			push = addv(push, rotatev(direction, 90, 0, 1, 0));
		}
		if (keyStates['d']) {
			push = addv(push, rotatev(direction, -90, 0, 1, 0));
		}

		// Makes the push vector the right length
		normalizev(&push);
		multiplyv(&push, ACCELERATION * 7);

		// Adds the push vector to the camera force and limits its speed
		if (!isSprinting && vlength(addv(cameraForce, push)) <= MAX_SPEED)
			cameraForce = addv(cameraForce, push);
		else if (!isSprinting && vlength(addv(cameraForce, push)) > MAX_SPEED && vlength(cameraForce) <= MAX_SPEED) {
			cameraForce = addv(cameraForce, push);
			normalizev(&cameraForce);
			multiplyv(&cameraForce, MAX_SPEED);
		}
		if (isSprinting && vlength(addv(cameraForce, push)) <= MAX_SPEED + SPRINT_BOOST)
			cameraForce = addv(cameraForce, push);
		else if (isSprinting && vlength(addv(cameraForce, push)) > MAX_SPEED + SPRINT_BOOST && vlength(cameraForce) <= MAX_SPEED + SPRINT_BOOST) {
			cameraForce = addv(cameraForce, push);
			normalizev(&cameraForce);
			multiplyv(&cameraForce, MAX_SPEED + SPRINT_BOOST);
		}

		// Adds the jump vector to the camera force
		//                                  GFORCE * (timeOfJump / 2 + 1)
		//                                               ||
		cameraForce = addv(cameraForce, createv(0, GFORCE * 16, 0));
	}
}

void motMoveCamera(void) 
{
	GLint v[4];
	glGetIntegerv(GL_VIEWPORT, v);
	// Moves mouse to the middle
	glutWarpPointer(v[2] / 2, v[3] / 2);

	// Checks whether to jump or not
	jumpFunc();
	
	// Aplys drag to the camera force if he isn't jumping
	if (!isJumping) {
		vector drag = cameraForce;
		normalizev(&drag);
		if (vlength(cameraForce) - DRAG > 0) {
			multiplyv(&drag, vlength(cameraForce) - DRAG);
		}
		else {
			drag = createv(0, 0, 0);
		}
		cameraForce = drag;
	}

	// Aplys air drag to the camera force if he is jumping
	else {
		vector drag = createv(cameraForce.x, 0, cameraForce.z);
		normalizev(&drag);
		if (vlength(createv(cameraForce.x, 0, cameraForce.z)) - AIR_DRAG > 0) {
			multiplyv(&drag, vlength(createv(cameraForce.x, 0, cameraForce.z)) - AIR_DRAG);
		}
		else {
			drag = createv(0, 0, 0);
		}
		cameraForce = createv(drag.x, cameraForce.y, drag.z);
	}

	// Makes the acceleration vector based on input and direction
	vector direction = substractv(createv(target.x, EYE_HEIGHT, target.z), eye);
	normalizev(&direction);
	vector acc = createv(0, 0, 0);
	if (keyStates['w']) {
		acc = addv(acc, direction);
	}
	if (keyStates['s']) {
		acc = substractv(acc, direction);
	}
	if (keyStates['a']) {
		acc = addv(acc, rotatev(direction, 90, 0, 1, 0));
	}
	if (keyStates['d']) {
		acc = addv(acc, rotatev(direction, -90, 0, 1, 0));
	}

	// Makes the acceleration vector the right length
	normalizev(&acc);
	multiplyv(&acc, ACCELERATION);

	// Makes the sprint vector
	vector sprint = acc;
	normalizev(&sprint);
	multiplyv(&sprint, 2 * ACCELERATION);

	// Adds the acceleration or the sprint vector to the camera force if he isn't jumping and limits its speed
	if (!isJumping) {
		if (!isSprinting && vlength(addv(cameraForce, acc)) <= MAX_SPEED)
			cameraForce = addv(cameraForce, acc);
		else if (!isSprinting && vlength(addv(cameraForce, acc)) > MAX_SPEED && vlength(cameraForce) <= MAX_SPEED) {
			cameraForce = addv(cameraForce, acc);
			normalizev(&cameraForce);
			multiplyv(&cameraForce, MAX_SPEED);
		}
		else if (!isSprinting && vlength(addv(cameraForce, acc)) > MAX_SPEED && vlength(cameraForce) > MAX_SPEED) {
			GLdouble length = vlength(cameraForce);
			cameraForce = addv(cameraForce, acc);
			normalizev(&cameraForce);
			multiplyv(&cameraForce, length);
		}
		if (isSprinting && vlength(addv(cameraForce, sprint)) <= MAX_SPEED + SPRINT_BOOST) 
			cameraForce = addv(cameraForce, sprint);
		else if (isSprinting && vlength(addv(cameraForce, sprint)) > MAX_SPEED + SPRINT_BOOST && vlength(cameraForce) <= MAX_SPEED + SPRINT_BOOST) {
			cameraForce = addv(cameraForce, sprint);
			normalizev(&cameraForce);
			multiplyv(&cameraForce, MAX_SPEED + SPRINT_BOOST);
		}
		else if (isSprinting && vlength(addv(cameraForce, sprint)) > MAX_SPEED + SPRINT_BOOST && vlength(cameraForce) > MAX_SPEED + SPRINT_BOOST) {
			GLdouble length = vlength(cameraForce);
			cameraForce = addv(cameraForce, sprint);
			normalizev(&cameraForce);
			multiplyv(&cameraForce, length);
		}
	}

	// Gravity
	cameraForce = addv(cameraForce, createv(0, - GFORCE, 0));

	// Check for floor
	if (eye.y + cameraForce.y < EYE_HEIGHT) {
		target.y += EYE_HEIGHT - eye.y;
		eye.y = EYE_HEIGHT;
		cameraForce = createv(cameraForce.x, 0, cameraForce.z);
		isJumping = false;
	}

	// Moves eye and target
	eye = addv(eye, cameraForce);
	target = addv(target, cameraForce);

	gluLookAt(eye.x, eye.y, eye.z, target.x, target.y, target.z, 0, 1, 0);
}

void freeCameraHandler (int x, int y) {
	GLint v[4];
	glGetIntegerv(GL_VIEWPORT, v); // Gets 4 values: X, Y position of window, Width and Heigth of viewport
	// Determines the angle on each axis based on mouse position
	cameraYaw += -45 + 90 * x / (GLdouble) v[2];
	cameraPitch += -30 + 60 * y / (GLdouble) v[3];

	// cameraYaw must not exeed 360 or be below -360 degrees
	cameraYaw -= ((int) cameraYaw / 360) * 360;

	// cameraPitch must not exeed 90 degrees or be below -90 degrees
	if (cameraPitch > 89) 
		cameraPitch = 89;
	if (cameraPitch < -89) 
		cameraPitch = -89;

	// Some notations
	GLdouble sinY = sin(-cameraPitch * DEG_TO_RAD);
	GLdouble sinX = sin(cameraYaw * DEG_TO_RAD);
	GLdouble cosX = cos(cameraYaw * DEG_TO_RAD);

	// Makes the target vector based on cameraYaw and cameraPitch (rotating the vector using rotatev() was a bit buggy)
	target = createv(cosX, 0, sinX);
	multiplyv(&target, sqrt(1 - sinY * sinY));
	target = addv(target, createv(0, sinY, 0));

	// Adds the eye position so that the camera points to the right place
	target = addv(target, eye);	
}

// Handles normal key presses (arrow, function and other keys not included)
void normalKeysHandler(unsigned char key, int x, int y) 
{
	switch(key) {
		case 'W':
		case 'w':
			keyStates['w'] = true;
			break;
		case 'S':
		case 's':
			keyStates['s'] = true;
			break;
		case 'A':
		case 'a':
			keyStates['a'] = true;
			break;
		case 'D':
		case 'd':
			keyStates['d'] = true;
			break;
		// Space Key
		case 32:
			keyStates[32] = true;
			break;
		// Esc Key
		case 27:
			exit(0);
			break;
	}
}

// Handles arrow, function and other keys presses (and maps them with normal keys)
void specialKeysHandler(int key, int x, int y) 
{
	switch(key) {
		case GLUT_KEY_UP:
			keyStates['w'] = true;
			break;
		case GLUT_KEY_DOWN:
			keyStates['s'] = true;
			break;
		case GLUT_KEY_LEFT:
			keyStates['a'] = true;
			break;
		case GLUT_KEY_RIGHT:
			keyStates['d'] = true;
			break;
		// Shift key
		case 112:
			isSprinting = true;
			break;
	}
}

// Handles normal keys releases (arrow, function and other keys not included)
void normalKeysUpHandler (unsigned char key, int x, int y) 
{
	switch(key) {
		case 'W':
		case 'w':
			keyStates['w'] = false;
			break;
		case 'S':
		case 's':
			keyStates['s'] = false;
			break;
		case 'A':
		case 'a':
			keyStates['a'] = false;
			break;
		case 'D':
		case 'd':
			keyStates['d'] = false;
			break;
		// Space Key
		case 32:
			keyStates[32] = false;
			break;
	}
}

// Handles arrow, function and other keys realeses (and maps them with normal keys releases)
void specialKeysUpHandler(int key, int x, int y) 
{
	switch(key) {
		case GLUT_KEY_UP:
			keyStates['w'] = false;
			break;
		case GLUT_KEY_DOWN:
			keyStates['s'] = false;
			break;
		case GLUT_KEY_LEFT:
			keyStates['a'] = false;
			break;
		case GLUT_KEY_RIGHT:
			keyStates['d'] = false;
			break;
		// Shift key
		case 112:
			isSprinting = false;
			break;
	}
}

void motionInit(void) {
	glutKeyboardFunc(normalKeysHandler);
	glutKeyboardUpFunc(normalKeysUpHandler);
	glutSpecialFunc(specialKeysHandler);
	glutSpecialUpFunc(specialKeysUpHandler);
	glutPassiveMotionFunc(freeCameraHandler);
	glutMotionFunc(freeCameraHandler);

	// Represents the camera position
	eye = createv(- 20, EYE_HEIGHT, 16);

	// Represents the point where the camera looks
	target = createv(0, EYE_HEIGHT, 0);

	// A vector that moves the camera
	cameraForce = createv(0, 0, 0);

	// Hides the mouse cursor
	glutSetCursor(GLUT_CURSOR_NONE);
}

vector motGetEyePos(void) {
	return eye;
}

vector motGetTargetPos(void) {
	return target;
}

vector motGetCamForce(void) {
	return cameraForce;
}

int motGetKeyStatus(int key) {
	return keyStates[key];
}

int motGetJump(void) {
	return isJumping;
}

int motGetSprint(void) {
	return isSprinting;
}